{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import ProtParam\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = \"files/sprot_human.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assignment(file_path, enzyme_rule, NTT, missed_cleavages, peptide_length):\n",
    "    \"\"\"\n",
    "    This function generates peptide substrings from protein sequences based on enzyme cleavage rules and other parameters.\n",
    "    \n",
    "    Args:\n",
    "        file_path (str): File path to the FASTA file that contains the protein sequences.\n",
    "        enzyme_rule (list): A list of amino acid residue/s that defines the cleavage rule of the enzyme used. \n",
    "                            This helps to filter out peptides that are unlikely to have been generated by the chosen enzyme, and thus reduces false positives\n",
    "        missed_cleavages (int): Number of missed cleavage sites allowed.  In some cases, where the enzyme may not cleave all of the possible peptide bonds due to various reasons, \n",
    "                            results in the formation of longer peptide fragments, which may be more difficult to analyze by MS. \n",
    "                            This parameter \"missed_cleavages\" is used to specify the maximum number of such uncleaved peptide bonds that are allowed during the enzymatic digestion process.\n",
    "        peptide_length (list): A list of two integers representing the minimum and maximum length of the generated peptide substrings.\n",
    "\n",
    "    Returns:\n",
    "        peptides (list): Thereafter returns a list of dictionaries containing peptide information such as the protein header and sequence,\n",
    "                        peptide number and sequence, peptide length, NTT, and missed cleavages -  which we can invoke for output\n",
    "        \n",
    "     \"\"\"\n",
    "    peptides = [] #initialize an empty list to store peptide information\n",
    "\n",
    "#loading the file and setting up the rules\n",
    "    with open(file_path) as handle:#open the FASTA file for reading\n",
    "        #use a loop to parse each sequence record in the FASTA file\n",
    "        for record in SeqIO.parse(handle, \"fasta\"): \n",
    "            prot_seq = str(record.seq) \n",
    "            prot_header = record.description\n",
    "\n",
    "            #iterate over all possible peptide sequences in the protein sequence\n",
    "            for i in range(len(prot_seq)):\n",
    "                for j in range(i + peptide_length[0], min(i + peptide_length[1] + 1, len(prot_seq))):\n",
    "                    peptide_seq = prot_seq[i:j]\n",
    "                    pep_len = len(peptide_seq)\n",
    "                    pep_NTT = 0\n",
    "\n",
    "                    #filter out peptides that don't meet the enzyme cleavage rules, peptide length, or NTT criteria\n",
    "                    if peptide_seq[0] not in enzyme_rule:\n",
    "                        continue\n",
    "                    if peptide_seq[-1] not in enzyme_rule:\n",
    "                        continue\n",
    "                    if pep_len < peptide_length[0] or pep_len > peptide_length[1]:\n",
    "                        continue\n",
    "                    for k in range(1, pep_len):\n",
    "                        if peptide_seq[k-1] in enzyme_rule and peptide_seq[k] not in enzyme_rule:\n",
    "                            pep_NTT += 1\n",
    "                    if pep_NTT > NTT:\n",
    "                        continue\n",
    "                    \n",
    "                    #iterate over all possible combinations of missed cleavage sites\n",
    "\n",
    "                    #For each missed cleavage site, we check if the resulting peptide sequence contains any residues that are not part of the enzyme rule, \n",
    "                    # and if it does not, then we add the peptide sequence to the list of peptides with the appropriate info such as protein header and sequence, \n",
    "                    # peptide number and sequence, peptide length, NTT, and missed cleavages.\n",
    "                    for k in range(1, missed_cleavages+2):\n",
    "                        for l in range(pep_len-k+1):\n",
    "                            missed_sites = [pos for pos, char in enumerate(peptide_seq) if char not in enzyme_rule]\n",
    "                            if l not in missed_sites:\n",
    "                                pep_missed_cleavages = k-1\n",
    "                                peptide_num = len(peptides) + 1\n",
    "                                pre_res = prot_seq[i+l-1] if i+l-1 >= 0 else \"-\"\n",
    "                                post_res = prot_seq[i+j] if i+j < len(prot_seq) else \"-\"\n",
    "                                peptides.append({\n",
    "                                    \"protein_header\": prot_header,\n",
    "                                    \"protein_seq\": prot_seq,\n",
    "                                    \"peptide_number\": peptide_num,\n",
    "                                    \"peptide_sequence\": peptide_seq,\n",
    "                                    \"pre_residue\": pre_res,\n",
    "                                    \"post_residue\": post_res,\n",
    "                                    \"peptide_length\": pep_len,\n",
    "                                    \"peptide_NTT\": pep_NTT,\n",
    "                                    \"peptide_missed_cleavages\": pep_missed_cleavages\n",
    "                                })\n",
    "    return peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = assignment(file_path, [['R', 'K'], 'C'], 2, 1, [7, 20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for result in results:\n",
    "    print(\"({}\\n,{}){}, {}, {}, {}\".format(\n",
    "        result['protein_header'],\n",
    "        result['protein_seq'],\n",
    "        result['peptide_sequence'][0],\n",
    "        result['peptide_sequence'][1:-1],\n",
    "        result['peptide_sequence'][-1],\n",
    "        result['peptide_length'],\n",
    "        result['peptide_missed_cleavages']\n",
    "    ))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
